### PR TITLE
[Windows] Fix crash swapping MainPage 

### LINKED ISSUE #7698: WinUI: Pushing the same page instance onto a NavigationPage after popping it causes a crash.
### Description

UWP only: 
Push a ContentPage onto a NavigationPage.
Pop the page, using the back button.
Push the same page again.
Exception = {"Catastrophic failure (0x8000FFFF (E_UNEXPECTED))"}

Minimal repro:
https://github.com/Keflon/MauiNavigationBugRepro

### Steps to Reproduce

1. Create a MAUI app.
2. Make this change in App.xaml.cs:
```csharp
public App()
{
    InitializeComponent();
    //MainPage = new AppShell();
    MainPage = new NavigationPage(new MainPage());
}
```
In MainPage.xaml.cs add the following, where `_childPage` is set to an instance of a Maui Page
```csharp
private void OnNavigateClicked(object sender, EventArgs e)
{
    App.Current.MainPage.Navigation.PushAsync(_childPage);
}
```
3. Add a Button in MainPage.xaml that raises the click handler.
```xaml
<Button 
    x:Name="NavigateBtn"
    Text="UWP Navigate Bug"
    SemanticProperties.Hint="Demonstrates a UWP navigation bug"
    Clicked="OnNavigateClicked"
    HorizontalOptions="Center" />
```

4. Click the Button.
5. Press the back button.
6. Click the button.
7. Notice UWP crashes with Exception = {"Catastrophic failure (0x8000FFFF (E_UNEXPECTED))"}

### Version with bug

6.0 (current)

### Last version that worked well

Release Candidate 3 (current)

### Affected platforms

Windows

### Affected platform versions

Latest

### Did you find any workaround?

No

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Fix crash swapping MainPage on Windows.

fix-7698

To validate the changes can use this sample  https://github.com/drasticactions/MauiRepros/tree/main/MauiReusePageBug from the issue or the added device test.

Tested the same on other platforms, and works on Android and iOS.

### Issues Fixed

Fixes #7698

### CODE REVIEW

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs [hunk 1]
```diff
@@ -7,13 +7,42 @@
...
+			var mainPage = new ContentPage
+			{
+				Content = button
+			};
+
+			var clicked = false;
+
+			button.Clicked += delegate
+			{
+				Application.Current.MainPage = mainPage;
```
Comment (PureWeen) [inline src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs:35]: For the test and I don't think we want to swap out the entire page of the test runner. 

If you look at the `MainPageSwapTests` tests, you'll see how we are standing up an Application/Window that's isolated ot the test

**Review by PureWeen — CHANGES_REQUESTED** [review]
The original issue was about pushing the same page a second time. 

If that's also broken and fixed by this PR then can you add a test for that as well?

**Review by mandel-macaque — APPROVED** [review]
+1 as long as @PureWeen concerns are addressed.

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs (hunk 1 above)
Comment (jsuarezruiz) [reply src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs]: Updated the tests.

File: src/Core/src/Handlers/Window/WindowHandler.Windows.cs [hunk 2]
```diff
@@ -61,6 +61,8 @@ public static void MapContent(IWindowHandler handler, IWindow window)
 			var previousRootView = windowManager.RootView;
 
 			windowManager.Disconnect();
+			handler.VirtualView.Content.Handler?.DisconnectHandler();
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Core/src/Handlers/Window/WindowHandler.Windows.cs:64]: I think the fix for this is to set the content of the `NavigationView` to null inside `NavigationRootManager` inside the `Disconnect` code

```C#
		public virtual void Disconnect()
		{
			_platformWindow.Activated -= OnWindowActivated;
			SetToolbar(null);
			if (_rootView.Content is RootNavigationView navView)
			{
				navView.Content = null;
			}

			_rootView.Content = null;
			_disconnected = true;
		}
```

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs [hunk 3]
```diff
@@ -3,17 +3,64 @@
...
+using Microsoft.UI.Xaml.Automation.Provider;
 using Xunit;
 using WPanel = Microsoft.UI.Xaml.Controls.Panel;
 
 namespace Microsoft.Maui.DeviceTests
 {
 	public partial class WindowTests : ControlsHandlerTestBase
 	{
+		[Fact(DisplayName = "Swapping MainPage no Crash")]
+		public async Task SwappingMainPageNoCrash()
```
Comment (PureWeen) [inline src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs:21]: I'm not sure if this test accurately tests the issue. 

I commented out all the changes in this PR and the test still passes.  One thing to note here is that the original issue was more about pushing the same page on the stack more then it was about swapping out the root page. 

So, the original issue is actually two different issues. 

1) is swapping out the rootpage https://github.com/drasticactions/MauiRepros/tree/main/MauiReusePageBug


2) is pushing a page, popping that page, pushing that page a second time.  https://github.com/Keflon/MauiNavigationBugRepro

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs [hunk 4]
```diff
@@ -171,5 +218,19 @@ public async Task MinimizeAndThenMaximizingWorks()
 				Assert.Equal(2, deactivated);
 			}
 		}
+
+		Task PerformClick(IButton button)
```
Comment (mattleibow) [inline src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs:222]: This should probably be in the shared assert extensions for reuse in other tests. Android has this feature as well, so we can do better test share.

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs [hunk 5]
```diff
@@ -3,17 +3,64 @@
...
+			};
+
+			var clicked = false;
+
+			button.Clicked += delegate
+			{
+				clicked = true;
+			};
+
+			await PerformClick(button);
```
Comment (mattleibow) [inline src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs:45]: Why does this need a button click? Can we just wait for the layout and then perform the navigation?

**Review by mattleibow — CHANGES_REQUESTED** [review]

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Core/src/Handlers/Window/WindowHandler.Windows.cs [hunk 6]
```diff
@@ -79,6 +79,8 @@ public static void MapContent(IWindowHandler handler, IWindow window)
...
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply src/Core/src/Handlers/Window/WindowHandler.Windows.cs]: Right, removed.

**Review by mattleibow — CHANGES_REQUESTED** [review]
I think Shane's comment on the test is probably the most critical - is the test correctly testing the 2 scenarios?

Other than that I was wondering if it were possible to reduce the complexity of the test by avoiding the clicks? The `CreateHandlerAndAddToWindow` already waits for a layout pass, so the code in the click could run inside that block. But is that click necessary? The code asserts if it is true, so it will stay true inside the `CreateHandlerAndAddToWindow` block.

If we remove the button then we can remove other code in the tests to make it simpler. And also there is a few whitespace changes that we can revert to simplify the PR some more.

**Review by mattleibow — CHANGES_REQUESTED** [review]
I think Shane's comment on the test is probably the most critical - is the test correctly testing the 2 scenarios?

Other than that I was wondering if it were possible to reduce the complexity of the test by avoiding the clicks? The `CreateHandlerAndAddToWindow` already waits for a layout pass, so the code in the click could run inside that block. But is that click necessary? The code asserts if it is true, so it will stay true inside the `CreateHandlerAndAddToWindow` block.

If we remove the button then we can remove other code in the tests to make it simpler. And also there is a few whitespace changes that we can revert to simplify the PR some more.

File: src/Core/src/Handlers/Window/WindowHandler.Windows.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply src/Core/src/Handlers/Window/WindowHandler.Windows.cs]: Changed.

File: src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs (hunk 4 above)
Comment (jsuarezruiz) [reply src/Controls/tests/DeviceTests/Elements/Window/WindowTests.Windows.cs:223]: Is not necessary to create the test. We can do the same without the Button.

File: src/Core/src/Handlers/Window/WindowHandler.Windows.cs [hunk 7]
```diff
@@ -1,4 +1,4 @@
-using System;
+﻿using System;
```
Comment (PureWeen) [inline src/Core/src/Handlers/Window/WindowHandler.Windows.cs:1]: undo changes on this file?

File: src/Core/src/Handlers/ContentView/ContentViewHandler.Windows.cs [hunk 8]
```diff
@@ -49,7 +49,9 @@ public static void MapContent(IContentViewHandler handler, IContentView page)
 		protected override void DisconnectHandler(ContentPanel platformView)
 		{
 			platformView.CrossPlatformLayout = null;
+      platformView.Children?.Clear();
```
Comment (PureWeen) [inline src/Core/src/Handlers/ContentView/ContentViewHandler.Windows.cs:52]: Fix tabs on this line?

**Review by PureWeen — CHANGES_REQUESTED** [review]

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Core/src/Handlers/Window/WindowHandler.Windows.cs (hunk 7 above)
Comment (jsuarezruiz) [reply src/Core/src/Handlers/Window/WindowHandler.Windows.cs]: Done.

**Review by PureWeen — APPROVED** [review]

**Review by mattleibow — APPROVED** [review]

**Pull request merged**